2020 Resource Hints chapter (#1587)
HTTPArchive/almanac.httparchive.org: src/content/en/2020/resource-hints.md

title: Resource Hints
description: Resource Hints chapter of the 2020 Web Almanac covering usage of dns-prefetch, preconnect, preload, prefetch, priority hints, and native lazy loading.
authors: [Zizzamia]

> reviewers: [jessnicolet, pmeenan, giopunt, mgechev, notwillk]

Reviewer: It looks like I'm in there (name is wrong).  Is up dating my name all I need to do there?

> reviewers: [jessnicolet, pmeenan, giopunt, mgechev, notwillk]

Reviewer 2: Ah yes I saw this:

![Resource Hints contributors](https://user-images.githubusercontent.com/10931297/100592497-153fb000-32ef-11eb-9793-ead32b6f76c0.png)

And presumed you were in there. If you could add your full name (and any other details you want to show in the Contributors page) that would be great.

analysts: [khempenius]
translators: []

> zizzamia_bio: Staff Software Engineer, Retail Growth at Coinbase

Author: Oh I wasn't aware was actually an option 😁

> zizzamia_bio: Staff Software Engineer, Retail Growth at Coinbase

Reviewer 3: It is! And by the way most authors usually write more about themselves than just title/employer. Try to answer the question: "Who are you and why are you qualified to write this chapter?"

> zizzamia_bio: Staff Software Engineer, Retail Growth at Coinbase

Reviewer 2: One thing to note is all the meta data is HTML-only. Really should add Markdown support as see some authors trying to use that, but that's for another days.

> zizzamia_bio: Staff Software Engineer, Retail Growth at Coinbase

Author: LOL, yeah I kind of put just my job as placeholder. I will give it a second pass in the next PR. thx @rviscomi

discuss: 2057
results: https://docs.google.com/spreadsheets/d/1lXjd8ogB7kYfG09eUdGYXUlrMjs4mq1Z7nNldQnvkVA/
queries: 21_Resource_Hints
#featured_quote: TODO @zizzamia
#featured_stat_1: TODO @zizzamia
#featured_stat_label_1: TODO @zizzamia
#featured_stat_2: TODO @zizzamia
#featured_stat_label_2: TODO @zizzamia
#featured_stat_3: TODO @zizzamia
#featured_stat_label_3: TODO @zizzamia
unedited: true
---

## Introduction

Over the past decade [Resource Hints](https://www.w3.org/TR/resource-hints/) have become essential primitives that allow developers to improve page performance and therefore the user experience. 

> Preloading resources and having browsers apply some intelligent prioritization is something that was actually started way back in 2009 by IE8 with something called the [preloader](https://speedcurve.com/blog/load-scripts-async/). In addition to the HTML parser, IE8 had a lightweight look-ahead preloader that scanned for tags that could initiate network requests (script, link, images).

Author: Right, I was referring to `<script>`, `<link>` and `<img>`. Great catch!

> Preloading resources and having browsers apply some intelligent prioritization is something that was actually started way back in 2009 by IE8 with something called the [preloader](https://speedcurve.com/blog/load-scripts-async/). In addition to the HTML parser, IE8 had a lightweight look-ahead preloader that scanned for tags that could initiate network requests (script, link, images).

Reviewer 2: ```suggestion
Preloading resources and having browsers apply some intelligent prioritization is something that was actually started way back in 2009 by IE8 with something called the [preloader](https://speedcurve.com/blog/load-scripts-async/). In addition to the HTML parser, IE8 had a lightweight look-ahead preloader that scanned for tags that could initiate network requests (`<script>`, `<link>`, and `<img>`).
```

- [Facebook.com](https://engineering.fb.com/2020/05/08/web/facebook-redesign/) used preload for faster navigation

Let’s take a look at most predominant Resource Hints supported by most browsers today: `dns-prefetch`, `preconnect`, `preload`,  `prefetch` and native lazy loading. 

When used with each individual hint we advise to always measure the impact before and after in the field, by using libraries like [WebVitals](https://github.com/GoogleChrome/web-vitals), [Perfume.js](https://github.com/zizzamia/perfume.js) or any other utility that supports the Web Vitals metrics. 

### dns-prefetch

[dns-prefetch](https://web.dev/preconnect-and-dns-prefetch/) helps resolve the IP address for a given domain ahead of time. As the [oldest](https://caniuse.com/link-rel-dns-prefetch) resource hint available, it uses minimal CPU and network resources, and helps the browser to avoid experiencing the "worst-case" delay for DNS resolution, which can be [over 1 second](https://www.chromium.org/developers/design-documents/dns-prefetching).

> Be mindful when using `dns-prefetch` as even if they are lightweight to do it's easy to exhaust browser limits for the number of concurrent in-flight DNS requests allowed (Chrome still has a [limit of 6](https://source.chromium.org/chromium/chromium/src/+/master:net/dns/host_resolver_manager.cc;l=353)).

Author: It might be more correct say "compared to `preconnect`" but we can keep it generic if you prefer. What you think?

> Be mindful when using `dns-prefetch` as even if they are lightweight to do it's easy to exhaust browser limits for the number of concurrent in-flight DNS requests allowed (Chrome still has a [limit of 6](https://source.chromium.org/chromium/chromium/src/+/master:net/dns/host_resolver_manager.cc;l=353)).

Reviewer 2: Yup that sounds good. Updated suggestion ready for you to commit if happy.

> ### preconnect

Author: Nice!

There are a variety of ways to implement the resources predictions logic, could be based on signals like user mouse movement, common user flows/journeys or even based on a combination of both on top of Machine Learning.

Be mindful, depending on the [quality](https://github.com/andydavies/http2-prioritization-issues#current-status) of HTTP/2 prioritization of the CDN used, `prefetch` prioritization could either improve performance or make it slower, by over prioritizing `prefetch` requests and taking away important bandwidth for the initial load. Make sure to double check the CDN you are using and adapt to take into consideration some of the best practices shared in [Andy Davies's](https://andydavies.me/blog/2020/07/08/rel-equals-prefetch-and-the-importance-of-effective-http-slash-2-prioritisation/) article.

### Native Lazy loading

> The [native lazy loading](https://web.dev/browser-level-image-lazy-loading/) hint is a native browser API for deferring the load of offscreen images and iframes. By using it, assets that are not needed during the initial page load will not initiate a network request, this will reduce data consumption and improve page performance.

Author: They sounds both correct to me, we might at this point just follow what the spec prefers https://html.spec.whatwg.org/multipage/urls-and-fetching.html#lazy-loading-attributes which is without hyphen.

> The [native lazy loading](https://web.dev/browser-level-image-lazy-loading/) hint is a native browser API for deferring the load of offscreen images and iframes. By using it, assets that are not needed during the initial page load will not initiate a network request, this will reduce data consumption and improve page performance.

Reviewer 2: Let's stick with lazy loading without the hypen then! Also means less to change form this year and last year's chapter :-)

> Be mindful Chromium's implementation of lazy-loading thresholds logic historically has been quite [conservative](https://web.dev/browser-level-image-lazy-loading/#distance-from-viewport-thresholds), keeping the offscreen limit to 3000px. During the last year the limit has been actively tested and improved on to better align developer expectations, and ultimately moving the thresholds to 1250px.

Author: Right right, def we should mention that!

## Resource Hints

Based on the HTTP Archive, let's jump into analyzing the 2020 trends, and compare the data with the previous 2019 dataset.

### Hints adoption

> More and more web pages are using the main Resource Hints, and in 2020 we are seeing the adoption remains incredibly consistent between desktop & mobile.

Author: I went back and forward on this sentence, at first I was leaning towards mentioning "mobile" got a slight larger percentage in adoption. Which to be fair I tend to user some resource hints more in "mobile", to gain some extra boost. BUT they made me notice the dataset between "mobile" and "desktop" is not similar enough to actually tell us a real difference between the two.

I will say let's drop "incredible" and just mention the adoption is similar between the two.

> More and more web pages are using the main Resource Hints, and in 2020 we are seeing the adoption remains incredibly consistent between desktop & mobile.

Reviewer 2: Sounds good to me.

```suggestion
More and more web pages are using the main Resource Hints, and in 2020 we are seeing the adoption remains consistent between desktop & mobile.
```

{{ figure_markup(
  image="TODO",
  caption="TODO",
  description="TODO",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTYAbLxN40s6mNR1jo0XDe_V4siN8TAsx2mryMp5IQmlJ-9O9eJxYROz7Rw6ozyFP6hlIZHxxh95GqX/pubchart?oid=1550112064&format=interactive",
  sheets_gid="1805612941",
  sql_file="adoption.sql"
) }}

> This is a great representation of how year over year we continue to focus on improving not only for desktop but also for mobile experiences. One take we will probably analyze even further next year is how HighEnd vs LowEnd devices will diverge in hints usage.

Author: Yeah, I feel the same. That's why I think next year we should focus a bit less of desktop vs mobile but more on HighEnd vs LowEnd. But for 2020 it is still a stretch for developer to dedicate time on mobile, imagine to dedicate time to LowEnd as well. LOL

Happy to rephrase this to just share the facts and remove any mobile vs desktop observation.

{{ figure_markup(
  caption="The percentage of desktop pages using DNS prefetching compared with other resource hints.",
  content="33%",
  classes="big-number",
  sheets_gid="1805612941",
  sql_file="adoption.sql"
) }}

The relative popularity of `dns-prefetch` is unsurprising as it first appeared in 2009, and has the widest support out of all major Resource Hints. It had a 4% increase in Desktop adoption, compared to [2019](https://almanac.httparchive.org/en/2019/resource-hints#resource-hints). We saw a similar increase for `preconnect` as well. One key reason this was the largest growth between all hints, is the clear and useful advice the Lighthouse audit is giving on this matter. Starting from this year's report we also introduce how the latest dataset performs against Lighthouse recommendations.

> `preload` usage has had a slower growth with only a 2% increase from 2019. This could be in part because it requires a bit more attention. While you only need the domain to use `dns-prefetch` and `preconnect`, you must specify the resource to use `preload`. While you can safely assume `dns-prefetch` and `preconnect` will improve performance, if `preload` is used on the wrong resource, it can actually damage performance.

Author: They do, only if used correctly. Let me see if I can find some articles even from the past with good benchmarks.

> `preload` usage has had a slower growth with only a 2% increase from 2019. This could be in part because it requires a bit more attention. While you only need the domain to use `dns-prefetch` and `preconnect`, you must specify the resource to use `preload`. While you can safely assume `dns-prefetch` and `preconnect` will improve performance, if `preload` is used on the wrong resource, it can actually damage performance.

Reviewer 2: Yeah my point is the "if used correctly". I think the current phrasing goes against what was said above.

What about this phrasing?:

```suggestion
`preload` usage has had a slower growth with only a 2% increase from 2019. This could be in part because it requires a bit more attention. While you only need the domain to use `dns-prefetch` and `preconnect`, you must specify the resource to use `preload`. While `dns-prefetch` and `preconnect` are reasonably low risk–though still can be abused– `preload` has a much greater potential to actually damage performance if used incorrectly.
```

### Hints per page

Across the board developers are learning how to better use Resource Hints, and compared to [2019](https://almanac.httparchive.org/en/2019/resource-hints#resource-hints) we've seen an improved use of `preload`, `prefetch` and `preconnect`. For expensive operations like preload and preconnect the median usage on desktop decreased from 2 to 1. We have seen the opposite for loading future resources with a lower priority with `prefetch`, with an increase from 1 to 2 in median per page.

{{ figure_markup(
  image="TODO",
  caption="TODO",
  description="TODO",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTYAbLxN40s6mNR1jo0XDe_V4siN8TAsx2mryMp5IQmlJ-9O9eJxYROz7Rw6ozyFP6hlIZHxxh95GqX/pubchart?oid=320451644&format=interactive",
  sheets_gid="175042082",
  sql_file="hints_per_page.sql"
) }}

> Resource hints are most effective when they're used selectively ("when everything is important, nothing is"). Having a more clear definition of what resources help improve Largest Contentful Paint moved the focus away from using `preconnect` and more towards `prefetch` by shifting some of the resource prioritization and freeing up bandwidth for what most helps the user at first.

Author: Yeah I can tell my point is not clear at all here. What I am trying to say it's less about the kind of resource but more about the timing when we make the prioritization. A developer that use `preload` for everything it could end up to damage LCP because we are over-prioritizing resources that are not important for LCP where instead we should move some of those late resources to a later stage by using `prefetch`.

I think mentioning `prefetch` here creates confusion, we can def drop it. But I still believe we should teach developers to be aware of the network timeline on when those resources are need it and how they impact LCP.

> Resource hints are most effective when they're used selectively ("when everything is important, nothing is"). Having a more clear definition of what resources help improve Largest Contentful Paint moved the focus away from using `preconnect` and more towards `prefetch` by shifting some of the resource prioritization and freeing up bandwidth for what most helps the user at first.

Reviewer 2: Oh OK. The specific use of LCP confused me. But you're using that to mean "initial rendering" I think?

What about something like this then?

```suggestion
Resource hints are most effective when they're used selectively ("when everything is important, nothing is"). Having a more clear definition of what resources help improve critical rendering, versus future navigation optimizations, can move the focus away from using `preconnect` and more towards `prefetch` by shifting some of the resource prioritization and freeing up bandwidth for what most helps the user at first.
```

{{ figure_markup(
  image="TODO",
  caption="TODO",
  description="TODO",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTYAbLxN40s6mNR1jo0XDe_V4siN8TAsx2mryMp5IQmlJ-9O9eJxYROz7Rw6ozyFP6hlIZHxxh95GqX/pubchart?oid=903180926&format=interactive",
  sheets_gid="1829901599",
  sql_file="as_attribute_by_year.sql"
) }}

Compared to the trend in 2019, we've seen rapid growth in font and style usage with the `as` attribute. This is likely related to both developers increasing the priority of critical CSS and also combining `preload` fonts with `display:optional` to [improve](https://web.dev/optimize-cls/#web-fonts-causing-foutfoit) Cumulative Layout Shift ([CLS](https://web.dev/cls/)).

Be mindful that omitting the as attribute, or having an invalid value will make it harder for the browser to determine the correct priority and in some cases, such as scripts, can even cause the resource to be fetched twice.

### The `crossorigin` attribute

> With `preload` and `preconnect` resources that have CORS enabled, such as fonts or images, it's important to include the `crossorigin` attribute, in order for the resource to be properly used. If the crossorigin attribute is absent, the request will follow the single-origin policy thereby making the use of preload useless.

Author: That's a great question, I will say generally no!
Based on [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) the images edge cases is more for:
- Images/video frames drawn to a canvas using drawImage().
- CSS Shapes from images.

> With `preload` and `preconnect` resources that have CORS enabled, such as fonts or images, it's important to include the `crossorigin` attribute, in order for the resource to be properly used. If the crossorigin attribute is absent, the request will follow the single-origin policy thereby making the use of preload useless.

Reviewer 2: I think as those are edge cases I wouldn't explicitly mention it and leave it as:

```
With `preload` and `preconnect` resources that have CORS enabled, such as fonts, it's important to include the `crossorigin` attribute, in order for the resource to be properly used. If the `crossorigin` attribute is absent, the request will follow the single-origin policy thereby making the use of preload useless.
```

You do say "such as" anyway so it's not like we're saying this is a definitive list.

The latest trend indicates that by using `preload` with the `crossorigin` attribute we have 34% anonymous (or equivalent) cases, and 0.43% use-credentials cases. This has evolved in conjunction with the increase in font-preloading mentioned earlier.

```html
<link rel="preload" href="ComicSans.woff2" as="font" type="font/woff2" crossorigin>
```

Be mindful that fonts preloaded without the `crossorigin` attribute will be fetched [twice](https://web.dev/preload-critical-assets/#how-to-implement-relpreload)!

### The `media` attribute

When it's time to choose a resource for use with different screen sizes, reach for the `media` attribute with `preload` to optimize your media queries.

```html
<link rel="preload" href="desktop.css" as="style" media=”only screen and (max-width: 768px)”>
<link rel="preload" href="mobile.css" as="style" media=”~~~”>
```

> The most popular media value used with `preload` is a combination of `only screen and (max-width: 768px)` and `screen and (min-width: 767px)`. Seeing over 2,100 different combinations of media queries in the 2020 dataset encourages us to consider how wide the variance is between concept and implementation of responsive design from site to site.

Reviewer 2: I'm not sure the most popular values are worth commenting on unless significantly more than the others? I was intrigued enough by this sentence [to look at the data](https://docs.google.com/spreadsheets/d/1lXjd8ogB7kYfG09eUdGYXUlrMjs4mq1Z7nNldQnvkVA/edit#gid=2016228012) and doesn't seem like these two values have any real significance:

![preload media values](https://user-images.githubusercontent.com/10931297/100557228-a2f3af00-329f-11eb-95d8-239138274bf2.png)

```suggestion
Seeing over 2,100 different combinations of media queries in the 2020 dataset encourages us to consider how wide the variance is between concept and implementation of responsive design from site to site.
```

Author: True, at the same time I find quite interesting the whole ecosystem is trying to conform on one particular group of number 768 and 767 to split desktop vs mobile.

I wonder how much the [breakpoints in w3schools](https://www.w3schools.com/css/css_rwd_mediaqueries.asp) contributed on this number.

What you think? Should we share something about this convergence or not?

Reviewer 2: I'd say it's more likely to be due to the popularity of Bootstrap than w3schools!

But yes I think you can say something about this. Maybe the second sentence rather than the first will de-emphasise it. Something like:

```suggestion
Seeing over 2,100 different combinations of media queries in the 2020 dataset encourages us to consider how wide the variance is between concept and implementation of responsive design from site to site. The ever popular `767px/768px` breakpoints (as popularised by Bootstrap amongst others) can be seen in the data.
```

{{ figure_markup(
  caption="Resource hints on desktop that use either preconnect or dns-prefetch use both in the same hint.",
  content="1.93%",
  classes="big-number",
  sheets_gid="281984550",
  sql_file="preconnect_and_dnsprefetch_relative.sql"
) }}

Close to 2% of pages (~40k) reported the issue of both `preconnect` & `dns-prefetch` in a single resource.

{{ figure_markup(
  caption="Pages that pass the preconnect Lighthouse audit",
  content="19.67%",
  classes="big-number",
  sheets_gid="152449420",
  sql_file="lighthouse_preconnect.sql"
) }}

We saw only 19.67% of pages passing Lighthouse’s "[Preconnect to required origins](https://web.dev/uses-rel-preconnect/)" audit, creating a large opportunity for thousands of websites to start using `preconnect` or `dns-prefetch` to establish early connections to important third-party origins.

{{ figure_markup(
  caption="Pages that pass the preload Lighthouse audit",
  content="84.6%",
  classes="big-number",
  sheets_gid="1047875076",
  sql_file="lighthouse_preload.sql"
) }}

> Running Lighthouse’s "[Preload key requests](https://web.dev/uses-rel-preload/)" audit resulted in 84.6% of pages passing the test, which is an astonishing result. If you are looking to use `preload` for the first time, remember, fonts and critical scripts are a good place to start.

Reviewer 2: I think it's important to understand what this audit means, as it's a bit less astonishing then.

It does NOT mean 84.6% of pages are using preload successfully. It means either there are, or there are no third level requests.

Crucially [this audit only looks at 1st party requests](https://github.com/GoogleChrome/lighthouse/blob/5c5de46ffce484f969ea5df8d71bce26dbc27ed6/lighthouse-core/audits/uses-rel-preload.js#L100) so things like Google Fonts aren't included.

Given that we only look at home pages are they less likely to have bundles apps with dependencies?

Think you should consider rephrasing this section.

### Native Lazy Loading

> Now let’s celebrate the first year of the [Native Lazy Loading](https://addyosmani.com/blog/lazy-loading/) API, which at the time of publishing already has over [72%](https://caniuse.com/loading-lazy-attr) browser support. This new API can be used to defer the load of below-the-fold iframes and images on the page until the user scrolls near them. This can reduce data usage, memory usage, and helps speed up above-the-fold content. Opting-in to lazy load is as simple as adding `loading=lazy`  on `<iframe>` or `<img>` elements.

Reviewer 2: Worth mentioning that Safari, as the last major browser to support this, has implemented this but not turned it on yet at which point support (for images at least) will jump into the high 90s?

Author: Do you have a link we could use to mention this information?

Reviewer 2: Can't find an official link (the [webkit bug](https://bugs.webkit.org/show_bug.cgi?id=196698) is marked as resolved

Let's just use caniuse: https://caniuse.com/loading-lazy-attr

Author: Sounds good!

{{ figure_markup(
  caption="Percentage of pages using native lazy loading set to “lazy”",
  content="3.87%",
  classes="big-number",
  sheets_gid="2039808014",
  sql_file="native_lazy_loading_attrs.sql"

> ) }}

Reviewer 2: Is this images or images and iframes? Think it would be good to clarify (applies to this whole section). And if just images should we have a sentence or two on iframes? We use it on our site :-)

Adoption is still in its early days, especially with the official thresholds earlier this year being too conservative, and only [recently](https://addyosmani.com/blog/better-image-lazy-loading-in-chrome/) aligning with developer expectations. With almost 72% of browsers supporting native image/source lazy loading, this is another area of opportunity especially for pages looking to improve data usage and performance on low-end devices. 

{{ figure_markup(
  caption="Percentage of pages passing the offscreen images Lighthouse audit",
  content="68.65%",
  classes="big-number",
  sheets_gid="1357389632",
  sql_file="lighthouse_offscreen_images.sql"
) }}

Running Lighthouse's "[Defer offscreen images](https://web.dev/offscreen-images/)" audit resulted in 68.65% of pages passing the test. For those pages there is an opportunity to lazy-load images after all critical resources have finished loading.

Be mindful to run the audit on both desktop and mobile as images may move off screen when the viewport changes.

> ## Machine Learning

Reviewer 2: What do you think about changing this title to predictive prefetching? As Machine Learning is just one way or predicting this.

```suggestion
## Predictive prefetching
```

Depending on the training set, the prefetching of Guess.js comes with over 90% accuracy.

Overall, predictive prefetching is still uncharted territory but combined with prefetching on mouse over and even a Service Worker strategy, it has great potential to provide instant experiences for all users of the website, while saving their data.

## HTTP/2 Push

HTTP/2 has a feature called server push that can potentially improve page performance when your product experiences long [RTTs](https://developer.mozilla.org/en-US/docs/Glossary/Round_Trip_Time_(RTT)) or server processing. In brief, rather than waiting for the client to send a request, the server preemptively pushes a resource that it predicts the client will request soon afterwards.

{{ figure_markup(
  caption="Percentage of HTTP/2 Push pages using preload/nopush",
  content="75.36%",
  classes="big-number",
  sheets_gid="308166349",
  sql_file="h2_preload_nopush.sql"
) }}

> In the 2020 dataset we have seen 1% of mobile pages using HTTP/2 Push, and of those 75% of preload header links use the nopush option in the firstHtml request. This means that even though a website is using preload, the majority of requests prefer to disable the push option.

Reviewer 2: I think it's important to explain the relationship between preload and push.

```suggestion
HTTP/2 Push is often initiated through the `preload` link header. In the 2020 dataset we have seen 1% of mobile pages using HTTP/2 Push, and of those 75% of preload header links use the `nopush` option in the page request. This means that even though a website is using the `preload` resource hint, the majority prefer to use just this and disable HTTP/2 pushing of that resource.
```

Author: Niiice, yeah that's much better! Thank you

> It's important to mention that HTTP/2 Push can also damage performance if not used correctly.

Reviewer 2: ```suggestion
It's important to mention that HTTP/2 Push can also damage performance if not used correctly which probably explains why is often disabled.
```

{{ figure_markup(
  image="TODO",
  caption="TODO",
  description="TODO",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTYAbLxN40s6mNR1jo0XDe_V4siN8TAsx2mryMp5IQmlJ-9O9eJxYROz7Rw6ozyFP6hlIZHxxh95GqX/pubchart?oid=252958553&format=interactive",
  sheets_gid="691299508",
  sql_file="adoption_service_workers.sql"
) }}

For `preload` on desktop we have an outstanding 47% rate of adoption and `prefetch` a 10% rate of adoption. In both cases the data is much higher compared to average adoption without a Service Worker.

As mentioned earlier, the [PRPL Pattern](https://addyosmani.com/blog/the-prpl-pattern/) will play a significant role in the coming years in how we combine Resource Hints with the Service Worker caching strategy.

## Future

> Let's dive into a couple of experimental hints. Very close to release we have Priority Hints, at the moment actively experimented with the web community. Lastly we have the 103 Early Hints in HTTP/2, which is still in early inception and there are a few players like Chrome and Fastly collaborating for upcoming test trials.

Reviewer 2: ```suggestion
Let's dive into a couple of experimental hints. Very close to release we have Priority Hints, at the moment actively experimented with the web community. Lastly we have the 103 Early Hints in HTTP/2, which is still in early inception and there are a few players like [Chrome and Fastly collaborating for upcoming test trials]](https://www.fastly.com/blog/beyond-server-push-experimenting-with-the-103-early-hints-status-code).
```

reviewers: [notwillk, giopunt, jessnicolet, pmeenan, mgechev]

> analysts: [khempenius]

Reviewer 3: Can you verify that all of these reviewers/analysts have meaningfully contributed to the chapter and are roughly in sorted order so that the biggest contributors are listed first?

### Native Lazy loading

The [native lazy loading](https://web.dev/browser-level-image-lazy-loading/) hint is a native browser API for deferring the load of offscreen images and iframes. By using it, assets that are not needed during the initial page load will not initiate a network request, this will reduce data consumption and improve page performance.

Be mindful Chromium's implementation of lazy-loading thresholds logic historically has been quite [conservative](https://web.dev/browser-level-image-lazy-loading/#distance-from-viewport-thresholds), keeping the offscreen limit to 3000px. During the last year the limit has been actively tested and improved on to better align developer expectations, and ultimately moving the thresholds to 1250px.

## Resource Hints

Based on the HTTP Archive, let's jump into analyzing the 2020 trends, and compare the data with the previous 2019 dataset.

### Hints adoption

More and more web pages are using the main Resource Hints, and in 2020 we are seeing the adoption remains incredibly consistent between desktop & mobile.

> // TODO @zizzamia Add Chart

Reviewer 3: I've published the chart to generate the embed URL and gotten the figure started but left out the image and caption/description.
```suggestion
{{ figure_markup(
  image="TODO",
  caption="TODO",
  description="TODO",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTYAbLxN40s6mNR1jo0XDe_V4siN8TAsx2mryMp5IQmlJ-9O9eJxYROz7Rw6ozyFP6hlIZHxxh95GqX/pubchart?oid=1550112064&format=interactive",
  sheets_gid="1805612941",
  sql_file="adoption.sql"
) }}
```

`prefetch` is used by 3% of sites on Desktop, making it the least widely used resource hint. This low usage may be explained by the fact that `prefetch` is useful for improving subsequent—rather than current—page loads. Thus, it will be overlooked if a site is only focused on improving its landing page, or the performance of the first page viewed. In the coming years with a more clear definition on what to measure for improving subsequent page experience, it could help teams prioritize `prefetch` adoption with clear performance quality goals to reach.

### Hints per page

Across the board developers are learning how to better use Resource Hints, and compared to [2019](https://almanac.httparchive.org/en/2019/resource-hints#resource-hints) we've seen an improved use of `preload`, `prefetch` and `preconnect`. For expensive operations like preload and preconnect the median usage on desktop decreased from 2 to 1. We have seen the opposite for loading future resources with a lower priority with `prefetch`, with an increase from 1 to 2 in median per page.

> // TODO @zizzamia Add chart 2

Reviewer 3: ```suggestion
{{ figure_markup(
  image="TODO",
  caption="TODO",
  description="TODO",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTYAbLxN40s6mNR1jo0XDe_V4siN8TAsx2mryMp5IQmlJ-9O9eJxYROz7Rw6ozyFP6hlIZHxxh95GqX/pubchart?oid=320451644&format=interactive",
  sheets_gid="175042082",
  sql_file="hints_per_page.sql"
) }}
```

Resource hints are most effective when they're used selectively ("when everything is important, nothing is"). Having a more clear definition of what resources help improve Largest Contentful Paint moved the focus away from using `preconnect` and more towards `prefetch` by shifting some of the resource prioritization and freeing up bandwidth for what most helps the user at first.

However, this hasn't stopped some misuse of the `preload` hint, since in one instance we discovered a page dynamically adding the hint and causing an infinite loop that created over 20k new preloads.

{{ figure_markup(
  caption="The most preload hints per page 🤯",
  content="20,931",
  classes="big-number",

>   sheets_gid="1805612941",

Reviewer 3: This gid value should correspond to the one in the figure above: https://docs.google.com/spreadsheets/d/1lXjd8ogB7kYfG09eUdGYXUlrMjs4mq1Z7nNldQnvkVA/edit#gid=175042082

If these values are just placeholders could you update the other figures to use TODOs instead?
```suggestion
  sheets_gid="175042082",
```

### The `as` attribute

With `preload` and `prefetch`, it's crucial to use the as attribute to help the browser prioritize the resource more accurately. Doing so allows for proper storage in the cache for future requests, applying the correct Content Security Policy ([CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP)), and setting the correct `Accept` request headers.

With `preload` many different content-types can be preloaded and the[full list](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/link#Attributes) follows the recommendations made in the Fetch [spec](https://fetch.spec.whatwg.org/#concept-request-destination). The most popular is the script `type` with 64% usage.

{{ figure_markup(
  caption="The percent of preload hints on mobile using the scripts type.",
  content="64%",
  classes="big-number",
  sheets_gid="1805612941",
  sql_file="// TODO @zizzamia Add sql"
) }}

This is likely related to a large group of sites built as Single Page Apps that need the main bundle as soon as possible to start downloading the rest of their JS dependencies. Subsequent usage comes from font at 8%, style at 5%, image at 1%, and fetch at 1%.

> // TODO @zizzamia Add chart 3

Reviewer 3: ```suggestion
{{ figure_markup(
  image="TODO",
  caption="TODO",
  description="TODO",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTYAbLxN40s6mNR1jo0XDe_V4siN8TAsx2mryMp5IQmlJ-9O9eJxYROz7Rw6ozyFP6hlIZHxxh95GqX/pubchart?oid=903180926&format=interactive",
  sheets_gid="1829901599",
  sql_file="as_attribute_by_year.sql"
) }}
```

It's important to mention that HTTP/2 Push can also damage performance if not used correctly.

One solution to this, is to use the [PRPL Pattern](https://addyosmani.com/blog/the-prpl-pattern/) which stands for **Push** (or preload) the critical resources, **Render** the initial route as soon as possible, **Pre-cache** remaining assets, and **Lazy-load** other routes and non-critical assets. This is possible only if your website is a Progressive Web App and uses a Service Worker to improve the caching strategy. By doing this all subsequent requests never even go out to the network, and so there's no need to push all the time and we still get the best of both worlds.

## Service Workers

For both `preload` and `prefetch` we've had an increase in adoption when the page is controlled by a [Service Worker](https://developers.google.com/web/fundamentals/primers/service-workers). This is because of the potential to both improve the resource prioritization by preloading when the Service Worker is not active yet and intelligently prefetching future resources while letting the Service Worker cache them before they're needed by the user.

> // TODO @zizzamia Add chart 4

Reviewer 3: ```suggestion
{{ figure_markup(
  image="TODO",
  caption="TODO",
  description="TODO",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vTYAbLxN40s6mNR1jo0XDe_V4siN8TAsx2mryMp5IQmlJ-9O9eJxYROz7Rw6ozyFP6hlIZHxxh95GqX/pubchart?oid=252958553&format=interactive",
  sheets_gid="691299508",
  sql_file="adoption_service_workers.sql"
) }}
```

This new hint can be used either as an HTML tag or by changing the priority of fetch requests via the `importance` option, which takes the same values as the HTML attribute.

With `preload` and `prefetch`, the priority is set by the browser depending on the type of resource. By using Priority Hints we can force the browser to change the default option.

{{ figure_markup(
  caption="The rate of priority hint adoption on mobile",
  content="0.77%",
  classes="big-number",
  sheets_gid="1805612941",
  sql_file="// TODO @zizzamia Add sql"
) }}

So far only 0.77% websites adopted this new hint as Chrome is still [actively](https://www.chromestatus.com/features/5273474901737472) experimenting, and at the time of this article&#39;s release the feature is on-hold.

{{ figure_markup(
  caption="Priority hints on mobile are on script elements",
  content="80%",
  classes="big-number",
  sheets_gid="1805612941",
  sql_file="// TODO @zizzamia Add sql"
) }}

The largest use is with script elements, which is unsurprising as the number of JS primary and third-party files continues to grow.

{{ figure_markup(

>   caption="Priority hints on mobile have &quot;low&quot; importance",

Reviewer 3: ```suggestion
  caption="Priority hints on mobile have \"low\" importance",
```

Author: Nice, great catch! I just added the fix requested.
